Add Multi-Profile Support

carmmmm opened the pull request:
# **Problem:**

The Synapse client currently only supports a single profile in its configuration file (`.synapseConfig`). This limitation prevents users from easily switching between multiple accounts or environments, which is a common need for users managing different credentials (e.g., personal vs. organizational accounts, testing vs. production environments).

****Affected Code Areas:****
- `synapseclient/core/credentials/credential_provider.py`
- `synapseclient/core/credentials/cred_data.py`
- `synapseclient/client.py`
- `synapseclient/api/configuration_services.py`

# **Solution:**

This PR introduces multi-profile authentication, similar to AWS CLI’s profile-based configuration. Users can now define multiple profiles in `.synapseConfig` and switch between them dynamically.

****Key Changes****
- Modified credential loading to support multiple profiles
- Updated the login() method to accept a profile argument
- Updated configuration parsing to retrieve the correct credentials for the selected profile

****New ~/.synapseConfig Format****
```
[default]
username = default_user
authtoken = default_auth_token

[profile user1]
username = user1
authtoken = user1_auth_token

[profile user2]
username = user2
authtoken = user2_auth_token
```

****How It Works****
Users can now log in using different profiles like this:

```
from synapseclient import Synapse

syn = Synapse()
syn.login(profile="user1")  # Logs in with user1's credentials
syn.login(profile="user2")  # Switches to user2's credentials
```

If no profile is specified, it defaults to [default].

# **Testing:**

****Manual Testing:****
1. Created multiple profiles in ~/.synapseConfig (see example above).
2. Tested logging in with different profiles using:
    ``` 
    syn.login(profile="user1")  # Verified correct username/token
    syn.login(profile="user2")  # Switched profiles and verified credentials
    ```
3. Printed debug statements to confirm correct credentials were loaded.
4. Checked default behavior (ensured [default] profile is used when no profile is provided).
5. Tested error handling:
      - Attempted login with a non-existent profile → correctly raised an error.
      - Verified no breaking changes for single-profile users.

****Next Steps****

- [x] Write unit tests to automate validation.
- [x] Update documentation in Synapse Python Client docs.
- [x] Cleanup debug p

rint statements before final merge.

****Feedback Request****
Before finalizing tests and documentation, I’d love your feedback:
- Does this implementation align with had in mind?
- Are there any additional edge cases that should be covered?
- The `.synapseConfig` file is manually created. Should there be a command to initialize profiles?
- The original `.synapseConfig` format used [authentication]. The new version separates profiles under [profile <name>]. This may cause issues with backward compatibility?


[SYNPY-893]: https://sagebionetworks.jira.com/browse/SYNPY-893

thomasyu888 reviewed (commented):
Thanks so much for your contribution @carmmmm !  We'll take a closer look soon at the inner workings, but high level.

Could you also add an example to this: https://github.com/Sage-Bionetworks/synapsePythonClient/blob/develop/synapseclient/.synapseConfig. for...

1. What a single profile configuration would look like in a multi-profile approach
2. What multiple profiles would look like

thomasyu888 commented on synapseclient/api/configuration_services.py:
```diff
@@ -76,23 +76,50 @@ def get_client_authenticated_s3_profile(
+        "authtoken": config[section].get("authtoken"),
+    }
+
+'''UNMODIFIED CODE!!!
```
In general, this isn't necessarily needed during code contribution, because you can always view the differences via pull requests.  This will reduce the amount of code you'll have to "delete" after review.

thomasyu888 commented on synapseclient/api/configuration_services.py:
```diff
@@ -76,23 +76,50 @@ def get_client_authenticated_s3_profile(
-def get_config_authentication(
-    config_path: str,
-) -> Dict[str, str]:
+''' CM MODIFED!!'''
```
The CM modified isn't necessary (but appreciated), because we can see your modifications in a PR.

thomasyu888 commented on synapseclient/client.py:
```diff
@@ -770,72 +818,147 @@ def setEndpoints(
 
+        normalized_profile = profile
+
+        # CM need to specify import from cred_data.py, still need to fix this!!
```
As a best practice, you can make use of 

```
TODO: need to specify import from cred_data.py
```

Different text editors like vscode will actually highlight these

carmmmm replied:
Done, thanks!

thomasyu888 commented on .gitignore:
```diff
@@ -32,3 +32,6 @@ coverage.xml
 .ipynb_checkpoints
 *.ipynb
+synapseclient/.synapseConfig
+synapseclient/.synapseConfig
+synapseclient/.synapseConfig
```
```suggestion
```

Remove duplicates here

thomasyu888 commented on synapseclient/api/configuration_services.py:
On our end - we should figure out how our tests can run on this PR. We have CI/CD running all the tests for things contributed.  Essentially, there's two ways to go about this.

1. Make this an additive change - where we support the multi profiles over time.
2. This is not backwards compatible - where we stop supporting the existing config. 

I'll defer to my team on Monday to decide what is the best way to go forward

BryanFauble replied:
I'll review more in this upcoming week, but I am in favor of "Make this an additive change - where we support the multi profiles over time.". We need to make sure this is a backwards compatible change. The change would look to see if the config file is in the new structure, and failing that, fall back to previous behavior.

The requirement to be backwards compatible is going to have a few implications with how the code in the PR was written.

Thanks for the contribution @carmmmm !

carmmmm replied:
Thank you @BryanFauble!

thomasyu888 commented on synapseclient/core/credentials/cred_data.py:
```diff
@@ -121,16 +123,63 @@ def __repr__(self):
+    def __init__(self, username: str = None, auth_token: str = None, profile: str = "default"):
+        self.username = username
+        self.auth_token = auth_token
+        self.profile = profile
```
A note for discussion: should MultiConfig be a separate class that then fills out this UserLoginArgs

BryanFauble replied:
I think it can probably be the same class between the two (See my comment here https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1183/files#r1999225453 on where I think the backwards compatibility needs to be added).

What I'd rather see is this class be converted over to a dataclass, similar to what I had created [for the project](https://github.com/Sage-Bionetworks/synapsePythonClient/blob/8761bab814f6c8c32cd99e8e2b3abde2243aef8d/synapseclient/models/project.py#L28-L178) (And other models)

We would create 3 attributes
1.  `username: Optional[str] = field(default=None)`
2. `auth_token: Optional[str] = field(default=None)` 
3. `profile: Optional[str] = field(default="default")`


We would we able to remove the custom `get` method you've added below since dataclasses automatically let you access the attributes on your class instance like:

```
my_login_args = UserLoginArgs(username="my_username", auth_token="my_token", profile="my_profile")
print(my_login_args.username)
print(my_login_args.auth_token)
print(my_login_args.profile)
```

Do we need to keep `to_dict`? Is it used anywhere.

`__repr__` dunder method would continue to remain since we are doing some sensitive data redaction there.

`get_config_authentication` would change per my comment: https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1183/files#r1999310788

carmmmm commented on synapseclient/client.py:
```diff
@@ -770,72 +818,147 @@ def setEndpoints(
             ),
         )
 
-        # Final check on login success
```
Got it! I’ll make sure to keep the older comments intact. Thanks for the heads-up!

carmmmm commented:
Thanks for the feedback @thomasyu888 ! I've added examples for both single-profile and multi-profile configurations in synapseclient/.synapseConfig. Let me know if there's anything else you'd like me to adjust! Also, I appreciate the improvements to the PR description—I reviewed the updates, and everything looks good on my end.

BryanFauble commented on synapseclient/.synapseConfig:
```diff
@@ -4,19 +4,24 @@
 
 ## Used for logging in to Synapse. See https://python-docs.synapse.org/tutorials/authentication/
```
We will need to make sure that we update this tutorial:

- https://python-docs.synapse.org/tutorials/authentication/

As well as consider how the CLI `synapse config` command will change as a result of these changes:

- https://python-docs.synapse.org/tutorials/command_line_client/#config

BryanFauble commented on synapseclient/.synapseConfig:
```diff
@@ -4,19 +4,24 @@
 #authtoken = <authtoken>
 
 
-## If you have projects with file stored on SFTP servers, you can specify your credentials here
-## You can specify multiple sftp credentials
-#[sftp://some.sftp.url.com]
-#username= <sftpuser>
-#password= <sftppwd>
-#[sftp://a.different.sftp.url.com]
-#username= <sftpuser>
-#password= <sftppwd>
```
SFTP config will need to remain as it's still a support feature of the client.

BryanFauble commented on synapseclient/api/configuration_services.py:
```diff
@@ -77,22 +77,31 @@ def get_client_authenticated_s3_profile(
-    return get_config_section_dict(
-        section_name=config_file_constants.AUTHENTICATION_SECTION_NAME,
-        config_path=config_path,
-    )
+    config = configparser.ConfigParser()
+    config.read(config_path)
+
+    section = f"profile {profile}" if profile != "default" else "default"
+
+    if section not in config:
+        raise ValueError(f"Profile '{profile}' not found in {config_path}")
+
+    return {
+        "username": config[section].get("username"),
+        "authtoken": config[section].get("authtoken"),
+    }
```
I think this is where we can make this a backwards compatible changes to support the new structure and fall back to the old one if it fails.

There is some extra exception handling occurring in `get_config_section_dict` that should continue to function as it does today. I would rather the logic here change to something like:



(This code isn't tested, but I think this is the logical code that should be here)
```python
section = f"profile {profile}" if profile != "default" else profile

section_for_profile = get_config_section_dict(
    section_name=section,
    config_path=config_path,
)

if section_for_profile:
    return section_for_profile

return get_config_section_dict(
    section_name=config_file_constants.AUTHENTICATION_SECTION_NAME,
    config_path=config_path,
)
```

BryanFauble commented on synapseclient/client.py:
```diff
@@ -216,26 +217,28 @@
-            syn = synapseclient.login(authToken="authtoken")
-
-        Using environment variable or `.synapseConfig`
+    Example:
+        syn = synapseclient.login(profile="user1")
```
For the `Example` block my expectation is that the provided code that may be copy/pasted by someone viewing the API docs. In this case we're missing the `import synapseclient` block.

```
    Example: Using a specific login profile
        When logging into Synapse you may specify a specific profile to use for the login credentials. In this case the `.synapseConfig` file is expected to contain a block that starts with `[profile user1]` and contains `username = your_username` and `authtoken = your_auth_token`

            import synapseclient
            syn = synapseclient.login(profile="user1")
```

I think we should also keep the other 2 examples assuming it continues to work:
`Logging in to Synapse using an authToken`, and `Using environment variable or `.synapseConfig`

The second example `Using environment variable or `.synapseConfig` could be expanded to note the behavior that it will now use the `default` block, or if not following the multi-profile approach, it'll continue to use the `authentication` block

BryanFauble commented on synapseclient/client.py:
```diff
@@ -400,15 +403,35 @@ def __init__(
         self.configPath = configPath
-        if os.path.isfile(configPath):
-            config = get_config_file(configPath)
-            if config.has_option("cache", "location"):
-                cache_root_dir = config.get("cache", "location")
-            if config.has_section("debug"):
-                config_debug = True
```
These 2 other options will need to remain in the code:

```
            if config.has_option("cache", "location"):
                cache_root_dir = config.get("cache", "location")
            if config.has_section("debug"):
                config_debug = True
```

BryanFauble commented on synapseclient/client.py:
```diff
@@ -400,15 +403,35 @@ def __init__(
+            config = get_config_file(self.configPath)
+
+            self.auth_profiles = {}
+            for section in config.sections():
+                if section.startswith("profile "):
+                    profile_name = section.split("profile ")[-1]
+                    self.auth_profiles[profile_name] = {
+                        "username": config.get(section, "username", fallback=None),
+                        "authtoken": config.get(section, "authtoken", fallback=None),
+                    }
+                elif section == "authentication":
+                    self.auth_profiles["default"] = {
+                        "username": config.get(section, "username", fallback=None),
+                        "authtoken": config.get(section, "authtoken", fallback=None),
+                    }
+        # If no profiles exist, create an empty one
+        if not hasattr(self, "auth_profiles") or not self.auth_profiles:
+            self.auth_profiles = {"default": {"username": None, "authtoken": None}}
+
+            '''OKAy bottom is og code, above to me whatever CM is edited 
+            '''
+            # if config.has_option("cache", "location"):
+            #     cache_root_dir = config.get("cache", "location")
+            # if config.has_section("debug"):
+            #     config_debug = True
```
I don't think this code is necessary is it? When a user calls the `syn.login(profile="user1")` method it performs a read of the given config file at that point in time. I t seems like in this case we do not need to do anything different in the `Synapse.__init__` function.

BryanFauble commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -104,48 +98,36 @@ def _create_synapse_credential(
 class UserArgsCredentialsProvider(SynapseCredentialsProvider):
+    def _get_auth_info(self, syn, user_login_args):
+
+        username = user_login_args.get("username")
+        token = user_login_args.get("auth_token")
+
+        if username and token:
+            return username, token
+
+        return None, None
     """
     Retrieves auth info from user_login_args during a CLI session.
     """
 
-    def _get_auth_info(
-        self, syn: "Synapse", user_login_args: Dict[str, str]
-    ) -> Tuple[str, str]:
-        return (
-            user_login_args.username,
-            user_login_args.auth_token,
-        )
-
-
```
If we change `UserLoginArgs` to be a dataclass there are some changes needed needed throughout this `credential_provider.py` script. Since everything would either be an instance of `UserLoginArgs`, or waiting to be converted into an instance of the `UserLoginArgs` dataclass.

BryanFauble reviewed (changes requested):
Thanks so much for the time to make these changes. Take a look at the comments I left and let me know what questions you have.

When you're ready for another round of review feel free to "re-request" review by clicking this button up here on the PR page:

![image](https://github.com/user-attachments/assets/96bea62e-2654-4e89-91e7-7785de8b99b3)

carmmmm commented:
Thank you so much for your feedback @BryanFauble! I've addressed creating the new data class, along with your other comments, and submitted re-request. Please let me know if this is what you had in mind, or any thing else to address. I appreciate it!

BryanFauble commented on synapseclient/api/configuration_services.py:
```diff
@@ -77,18 +77,32 @@ def get_client_authenticated_s3_profile(
-def get_config_authentication(
-    config_path: str,
-) -> Dict[str, str]:
+def get_config_authentication(config_path: str, profile: str = "default"):
```
We should keep typing checking for the return types in place

```suggestion
def get_config_authentication(config_path: str, profile: str = "default")  -> Dict[str, str]:
```

BryanFauble commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -104,47 +94,39 @@ def _create_synapse_credential(
 
 
 class UserArgsCredentialsProvider(SynapseCredentialsProvider):
+    def _get_auth_info(self, syn, user_login_args):
+
+        username = user_login_args.username
+        token = user_login_args.auth_token
+        # username = user_login_args.get("username")
+        # token = user_login_args.get("auth_token")
+
+        if username and token:
+            return username, token
+
+        return None, None
```
It's valid to have a token without specifying your username.

We should also look to have types in place for the function arguments and return type:
`def _get_auth_info(self, syn, user_login_args):`

BryanFauble commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -104,47 +94,39 @@ def _create_synapse_credential(
+            return None, user_login_args.auth_token  # No username needed, just authToken
 
-        return username, token
+        # Otherwise, fall back to profile-based lookup
+        if not user_login_args.profile:
+            raise SynapseAuthenticationError("No authentication method provided (neither authToken nor profile).")
+
+        raise SynapseAuthenticationError(f"Profile '{user_login_args.profile}' not found in {syn.configPath}")
```
I'm not sure that this is correct. You have 2 branches here:

1. `if not user_login_args.profile:` -> An exception will be raised.
2. If ^ does not evaluate to True an exception is still raised.

BryanFauble commented on synapseclient/core/credentials/cred_data.py:
```diff
@@ -121,16 +125,87 @@ def __repr__(self):
+    username: Optional[str] = field(default=None)
+    auth_token: Optional[str] = field(default=None, repr=False)  # Hide auth_token from debug logs
+
+    def __post_init__(self):
+        """
+        Automatically fetch credentials from the configuration file if none are provided.
+
+        This method ensures that if a username and auth token are not explicitly supplied,
+        they will be retrieved from the `~/.synapseConfig` file based on the specified profile.
+
+        Raises:
+            ValueError: If the specified profile is not found in the configuration file.
+        """
+        if self.auth_token is None and self.username is None:
+            auth_profiles = get_config_authentication()
+            if self.profile in auth_profiles:
+                self.username = auth_profiles[self.profile]["username"]
+                self.auth_token = auth_profiles[self.profile]["auth_token"]
+            else:
+                raise ValueError(f"Profile '{self.profile}' not found in config.")
```
I do not think that we should preemptively retrieve this data. I see that the only purpose of this `UserLoginArgs` is to hold it's 3 attributes in one place.

It's usage of `get_config_authentication` is also a concern. There are now 2 instances of this function, whereas we should only have 1.

`synapseclient/core/credentials/cred_data.py::get_config_authentication` should be removed.

`synapseclient/api/configuration_services.py::get_config_authentication` Should be the one that is kept and re-used wherever we need to look the data up.


`synapseclient/api/configuration_services.py::get_config_authentication` should only be called at the point in time when the data is needed - Not ahead of time as done here.

BryanFauble reviewed (changes requested):
Thanks again for another round!

carmmmm commented:
Thank you!

BryanFauble commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -52,28 +59,20 @@ def _get_auth_info(
-        Arguments:
-            syn: Synapse client instance
-            user_login_args: subset of arguments passed during syn.login()
+        if not user_login_args.auth_token and not user_login_args.username:
+            auth_profiles = get_config_authentication(syn.configPath, user_login_args.profile)
+            user_login_args.username = auth_profiles["username"]
+            user_login_args.auth_token = auth_profiles["auth_token"]
```
Te should probably only check for `if not user_login_args.auth_token`.

And then only set the `user_login_args.username`, `if not user_login_args.username`

BryanFauble commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -104,47 +103,66 @@ def _create_synapse_credential(
+        """
+
+        token: Optional[str] = user_login_args.auth_token
+        username: Optional[str] = user_login_args.username if user_login_args.username else None
```
This ternary statement shouldn't be needed. We should be able to just do `username: Optional[str] = user_login_args.username`

BryanFauble commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -104,47 +103,66 @@ def _create_synapse_credential(
-        username = config_dict.get("username")
-        token = config_dict.get("authtoken")
-
-        if user_login_args.username and username != user_login_args.username:
-            # if the username is provided and there is a config file username but they
-            # don't match then we don't use any of the values from the config
-            # to prevent ambiguity
-            username = None
-            token = None
-            syn.logger.warning(
-                f"{user_login_args.username} was defined in the user login "
-                "arguments, however, it is also defined in the `~/.synapseConfig` "
-                "file. Becuase they do not match we will not use the `authtoken` "
-                "in the `~/.synapseConfig` file.",
-            )
```
This username check should continue to exist

BryanFauble commented on synapseclient/api/configuration_services.py:
```diff
@@ -77,18 +77,32 @@ def get_client_authenticated_s3_profile(
-        The authentication section of the configuration file
+        dict: A dictionary with 'username' and 'authtoken'.
     """
+    config = configparser.ConfigParser()
+    config.read(config_path)
```
^

BryanFauble commented on synapseclient/api/configuration_services.py:
```diff
@@ -76,24 +77,46 @@ def get_client_authenticated_s3_profile(
+    config = configparser.ConfigParser()
+    config.read(config_path)
+
+    auth_profiles = {}
+
+    # Extract all profiles and normalize them
+    for section in config.sections():
+        normalized_profile = section.replace("profile ", "").strip()  # Ensure no trailing spaces
+        auth_profiles[normalized_profile] = {
+            "username": config.get(section, "username", fallback=None),
+            "auth_token": config.get(section, "authtoken", fallback=None),
+        }
+
+    # **Check for the exact match**
+    if profile in auth_profiles:
+        return auth_profiles[profile]
 
+    # If profile not found, return error instead of "default"
+    raise SynapseNoCredentialsError(f"Profile '{profile}' not found in the config file.")
```
The code you're doing here isn't needed and we shouldn't be modifying potential behavior. See my comment I left: https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1183#discussion_r1999225453 on what I think the code should be instead.

BryanFauble commented on synapseclient/client.py:
```diff
@@ -400,16 +402,19 @@ def __init__(
             if config.has_section("debug"):
                 config_debug = True
 
+            self.auth_profiles = {}
+
```
This doesn't need to be an attribute on the `Synapse` class

BryanFauble commented on synapseclient/client.py:
```diff
@@ -771,70 +776,64 @@ def setEndpoints(
+            syn = Synapse() # Instantiate once
+            syn.login(authToken="your_auth_token")
         """
-        # Note: the order of the logic below reflects the ordering in the docstring above.
-
-        # Check version before logging in
-        if not self.skip_checks:
-            version_check()
-
-        # Make sure to invalidate the existing session
```
Code and comments shouldn't be removed

BryanFauble commented on synapseclient/client.py:
```diff
@@ -771,70 +776,64 @@ def setEndpoints(
+        # Ensure previous session is cleared
         self.logout()
 
+        user_login_args = UserLoginArgs(profile=profile or "default", username=email, auth_token=authToken)
```
```suggestion
        user_login_args = UserLoginArgs(profile=profile, username=email, auth_token=authToken)
```

You are already setting a default for `profile` in the argument of the function: `profile: str = "default"`

BryanFauble commented on synapseclient/client.py:
```diff
@@ -771,70 +776,64 @@ def setEndpoints(
         if not silent:
             display_name = self.credentials.displayname or self.credentials.username
-            self.logger.info(f"Welcome, {display_name}!\n")
+            self.logger.info(f"Welcome, {display_name}!(Profile: {profile})\n")
```
I'm not sure I like the format of the message, but I'm not certain what would look great here.

My other concern is in the context of a user who hasn't moved into the multi-profile approach, this isn't really an applicable message for them. Since they wouldn't be using a `default` or specific profile.

carmmmm replied:
Hm those are good points, right now for a user using the one profile it reads `Welcome, Username!(Profile: authentication)' which is a bit clunky I agree. I've modified it so that it reads clearer, and only shows the profile on users with a multi profile configuration. Let me know if that's better! If not happy to pivot. 
```
        if not silent:
            display_name = self.credentials.displayname or self.credentials.username
            if profile.lower() == "authentication":
                self.logger.info(f"Welcome, {display_name}!\n")
            else:
                self.logger.info(f"Welcome, {display_name}! You are using the '{profile}' profile.")
```

BryanFauble replied:
I like the suggestion! The only comment is that instead of using a string literal here `if profile.lower() == "authentication":`, use the constant available with `config_file_constants.AUTHENTICATION_SECTION_NAME`

`if profile.lower() == config_file_constants.AUTHENTICATION_SECTION_NAME:`

BryanFauble commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -52,28 +59,20 @@ def _get_auth_info(
-        return self._create_synapse_credential(
-            syn, *self._get_auth_info(syn=syn, user_login_args=user_login_args)
-        )
+        return self._create_synapse_credential(syn, user_login_args.username, user_login_args.auth_token)
```
`self._get_auth_info` must be called in the final solution.

`_get_auth_info` is decorated with `@abc.abstractmethod` meaning anything using this `SynapseCredentialsProvider` base class must implement that method.

Look at `class AWSParameterStoreCredentialsProvider(SynapseCredentialsProvider):` for example. Your suggested changes will break that as a valid login method.

carmmmm replied:
Thanks for the feedback! I made sure _get_auth_info() is always called before falling back to .synapseConfig, so subclasses like AWSParameterStoreCredentialsProvider should still work as expected. Now, _get_auth_info() handles retrieving credentials first, and only if there’s no valid auth token, it checks .synapseConfig. Would you mind confirming that this approach aligns with the intended behavior for all credential providers, and if the approach looks good?

```
    def get_synapse_credentials(
        self, syn: "Synapse", user_login_args: UserLoginArgs
    ) -> Union[SynapseCredentials, None]:
        """
        Returns
        [SynapseCredentials][synapseclient.core.credentials.cred_data.SynapseCredentials]
        if this provider is able to get valid credentials, returns None otherwise.

        Arguments:
            syn: Synapse client instance
            user_login_args: subset of arguments passed during syn.login()

        Returns:
            [SynapseCredentials][synapseclient.core.credentials.cred_data.SynapseCredentials]
                if valid credentials can be found by this provider, None otherwise.
        """
        # Retrieve authentication info using the appropriate method
        username, auth_token = self._get_auth_info(syn=syn, user_login_args=user_login_args)

        # If authentication token is not provided, fall back to configuration file
        if not auth_token:
            auth_profiles = get_config_authentication(syn.configPath, user_login_args.profile)
            if not username:
                username = auth_profiles["username"]
            auth_token = auth_profiles["authtoken"]

        return self._create_synapse_credential(syn, username, auth_token)

```

BryanFauble commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -267,7 +275,7 @@ def get_credentials(
         UserArgsCredentialsProvider(),
         ConfigFileCredentialsProvider(),
         EnvironmentVariableCredentialsProvider(),
-        AWSParameterStoreCredentialsProvider(),  # see service catalog issue: SC-260
+        AWSParameterStoreCredentialsProvider(),
```
Please leave comments like this around

BryanFauble commented on Pipfile:
```diff
@@ -5,6 +5,7 @@ verify_ssl = true
 
 [packages]
 synapseclient = {file = ".", editable = true, path = "."}
+config = "*"
```
Whats this updated `packages` section for? This change should not require any updated dependencies or changes to the install process.

carmmmm replied:
That change wasn't intentional—looks like Pipfile got updated when I ran pipenv install. I'll revert it to ensure no unnecessary dependencies are added. Sorry I missed that!

BryanFauble commented on synapseclient/.synapseConfig:
```diff
@@ -4,10 +4,23 @@
 
 ## Used for logging in to Synapse. See https://python-docs.synapse.org/tutorials/authentication/
 ## for information on retrieving an auth token.
+## Single-profile configuration
 #[authentication]
 #username = <username>
 #authtoken = <authtoken>
```
We can remove these lines. In this PR we will still support people using this `[authentication]` structure, but we should not promote it's usage. Instead, let's only promote the usage of the `## Multi-profile configuration` way of doing things.

BryanFauble commented on synapseclient/api/configuration_services.py:
```diff
@@ -76,24 +77,46 @@ def get_client_authenticated_s3_profile(
+        FileNotFoundError: If the config file does not exist.
+
+    Example:
+        >>> get_config_authentication("~/.synapseConfig", profile="user1")
+        {'username': 'user1', 'auth_token': 'user1_auth_token'}
```
This example could use an update. Ideally I should be able to copy/paste the example into a totally new python script and be able to run it.

BryanFauble reviewed (approved):
Thanks for another round. Please feel free to go back and mark conversations are "Resolved" once you've addressed the item in question.

BryanFauble reviewed (changes requested):
Thanks for another round. Please feel free to go back and mark conversations are "Resolved" once you've addressed the item in question.

BryanFauble commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -109,28 +118,52 @@ class UserArgsCredentialsProvider(SynapseCredentialsProvider):
+            if the specified profile does not exist in the configuration file.
+        """
+        # check to make sure we didn't accidentally provide the wrong user
+        config_dict = get_config_authentication(config_path=syn.configPath)
```
```suggestion
        config_dict = get_config_authentication(config_path=syn.configPath, profile=user_login_args.profile)
```

BryanFauble commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -109,28 +118,52 @@ class UserArgsCredentialsProvider(SynapseCredentialsProvider):
+        """
+        # check to make sure we didn't accidentally provide the wrong user
+        config_dict = get_config_authentication(config_path=syn.configPath)
+        username = config_dict.get("username")
```
Flagging this section to look back at the logic tomorrow.

BryanFauble commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -67,9 +68,17 @@ def get_synapse_credentials(
         """
-        return self._create_synapse_credential(
-            syn, *self._get_auth_info(syn=syn, user_login_args=user_login_args)
-        )
+        # Retrieve authentication info using the appropriate method
+        username, auth_token = self._get_auth_info(syn=syn, user_login_args=user_login_args)
+
+        # If authentication token is not provided, fall back to configuration file
+        if not auth_token:
+            auth_profiles = get_config_authentication(syn.configPath, user_login_args.profile)
+            if not username:
+                username = auth_profiles["username"]
+            auth_token = auth_profiles["authtoken"]
+
+        return self._create_synapse_credential(syn, username, auth_token)
```
I don't think that we can fall back to using the configuration file in this base class, since using the config file is specific only to the parent class `ConfigFileCredentialsProvider`.

Rather, in this base class I think that we would continue with the logic that was already here, and then in any classes that are implementing this base class - Handle for any specific logic required.

```
        return self._create_synapse_credential(
            syn, *self._get_auth_info(syn=syn, user_login_args=user_login_args)
        )
```

BryanFauble commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -144,7 +177,24 @@ def _get_auth_info(
                 "in the `~/.synapseConfig` file.",
             )
 
-        return username, token
+        # If authToken is explicitly provided, return it directly and skip profiles
+        if user_login_args.auth_token:
+            return user_login_args.username, user_login_args.auth_token
+
+        # Otherwise, fall back to profile-based lookup
+        if not user_login_args.profile:
+            raise SynapseAuthenticationError("No authentication method provided (neither authToken nor profile).")
+
+        # Fetch available profiles from the config file
+        auth_profiles = get_config_authentication(syn.configPath)
+
+        # If the profile exists, return its credentials
+        if user_login_args.profile in auth_profiles:
+            profile_data = auth_profiles[user_login_args.profile]
+            return profile_data["username"], profile_data["auth_token"]
+
+        # Otherwise, raise an error
+        raise SynapseAuthenticationError(f"Profile '{user_login_args.profile}' not found in {syn.configPath}")
```
This logic shouldn't be executed. At this point in time after calling:

`config_dict = get_config_authentication(config_path=syn.configPath, profile=user_login_args.profile)`

`config_dict` is going to contain either a filled in dictionary with data for the profile the user is logging in as, OR, a blank dictionary. Both as acceptable returns in this context because credentials are retrieved in a chain:

```
DEFAULT_CREDENTIAL_PROVIDER_CHAIN = SynapseCredentialsProviderChain(
    cred_providers=[
        UserArgsCredentialsProvider(),
        ConfigFileCredentialsProvider(),
        EnvironmentVariableCredentialsProvider(),
        AWSParameterStoreCredentialsProvider(),  # see service catalog issue: SC-260
    ]
)
```


When `ConfigFileCredentialsProvider` fails (This method here) we do not want to stop from progressing to the next provider in the chain (`EnvironmentVariableCredentialsProvider`). When this method fails to return information we don't need to log or note anything, because it may be the case the user never intended to use credentials from the config file.


We could however, provide a warning message in this method here, like:
```
        if user_login_args.profile and user_login_args.profile.lower() != "default":
            syn.logger.warning(f"The profile ({user_login_args.profile}) was provided while logging in, but could not be found in your `~/.synapseConfig` file. See https://python-docs.synapse.org/tutorials/authentication/ for more details on authentication.")
```

BryanFauble reviewed (changes requested):
I left a few more comments, thanks again! You're on the last stretch as far as this logic goes!

From here the last few checklist items that I see are:

- [ ] Updating the logic for the `synapse config` [CLI command ](https://python-docs.synapse.org/tutorials/command_line_client/#config)to support creating the updated multi-profile config file along with profiles, or a default
- [ ] Updating the logic for the `synapse login` [CLI command](https://python-docs.synapse.org/tutorials/command_line_client/#login) to support a profile flag.
- [ ] Updating [this Authentication document](https://python-docs.synapse.org/tutorials/authentication/) to describe our support for multiple profiles
- [ ] Writing some unit tests for this logic (Integration tests are required in this context since data doesn't need to be sent to Synapse to verify this logic works) for users using this as a command line client, and for users using this as a python library to write code

carmmmm commented:
I believe this addresses all the remaining items you mentioned:

✅ Updated the synapse config CLI logic to support creating a multi-profile config file, including support for named profiles and default.
✅ Updated the synapse login CLI to support a --profile flag.
✅ Updated the [Authentication doc](https://github.com/Sage-Bionetworks/synapsePythonClient/blob/develop/docs/authentication.rst) to describe the new multi-profile functionality.
✅ Added unit tests for both CLI and Python usage (non-integration, as no Synapse communication is required).

Let me know if you spot anything else I missed! Otherwise, excited to hear your thoughts — and thank you again for the guidance along the way 😊

BryanFauble commented on docs/tutorials/authentication.md:
```diff
@@ -33,47 +34,84 @@ Welcome, First Last!
 The following describes how to add your credentials to the `~/.synapseConfig` file without the use of the `synapse config` command.
 
-Open the `~/.synapseConfig` file and find the following section:
+Open the `~/.synapseConfig` file (can use 'nano ~/.synapseConfig' to edit or 'cat ~/.synapseConfig' to read from the CLI)
+and find the following section:
```
```suggestion
Open the `~/.synapseConfig` file using your preferred text editing tool and find the following section:
```

BryanFauble commented on synapseclient/__main__.py:
```diff
@@ -534,21 +535,27 @@ def _replace_existing_config(path, auth_section):
+        )
+
+    if profile_name:
+        profile_section = f"[{profile_name}]\n{auth_section.strip()}\n"
```
```suggestion
        profile_section = f"[profile {profile_name}]\n{auth_section.strip()}\n"
```

BryanFauble commented on synapseclient/__main__.py:
```diff
@@ -564,15 +571,16 @@ def config(args, syn):
 
-    auth_section = "[authentication]\n"
+    profile_name = args.profile or "default"
+    auth_section = f"[{profile_name}]\n"
```
This would need to conditionally use `profile` as the prefix if profile_name is not default

carmmmm replied:
Thanks! You're right, [default] should remain unprefixed, and named profiles should use the profile prefix. I'll update the logic so the section header becomes:

```
section_header = f"[{profile_name}]\n" if profile_name == "default" else f"[profile {profile_name}]\n"
auth_section = section_header + auth_section_content
```
That should ensure the correct format based on the profile. Thanks again!

carmmmm commented:
Thanks so much for the review! I’ve gone ahead and addressed the feedback and pushed the updates. Let me know if anything else comes up during manual testing or mkdocs review — happy to keep iterating!

linglp commented on docs/tutorials/authentication.md:
````diff
@@ -19,6 +19,7 @@ Use the [synapseclient.login][synapseclient.Synapse.login] function
 ```python
 import synapseclient
 syn = synapseclient.login(authToken="authtoken")
+#returns Welcome, First Last! 
````
nit: is this something meant to be deleted?

BWMac replied:
I think this is a note for what should be printed back to the user if they have succeeded in authenticating. This shows up in the mkdocs site and I think its pretty helpful.

linglp commented on docs/tutorials/authentication.md:
```diff
@@ -33,47 +34,83 @@ Welcome, First Last!
 
+The Synapse Python Client supports multiple profiles within the ~/.synapseConfig file, enabling users to manage credentials for multiple accounts.
+
+Each profile is defined in its own '[profile <profile_name>]' section. A default profile can still be defined using '[default]'.
```
nit: I think it might make sense to just use backtick `[profile <profile_name>]`

carmmmm replied:
Thanks for the suggestion! I updated the formatting with backticks for clarity.

linglp commented on docs/tutorials/authentication.md:
````diff
@@ -33,47 +34,83 @@ Welcome, First Last!
 ```
-> synapse config
-Synapse username (Optional):
+#For Default profile
````
To make this more clear, maybe you could say something like:
"for modifying the username used in the default profile"

linglp commented on docs/tutorials/authentication.md:
```diff
@@ -33,47 +34,83 @@ Welcome, First Last!
+
+Auth token: $MY_SYNAPSE_TOKEN
+
+#For additional profiles
```
similarly, I think here you could say something like: "for adding a new profile with a new username"

linglp commented on synapseclient/.synapseConfig:
```diff
@@ -4,10 +4,18 @@
 
 ## Used for logging in to Synapse. See https://python-docs.synapse.org/tutorials/authentication/
 ## for information on retrieving an auth token.
```
It looks like that if an user has an outdated .synapseConfig, this will be a breaking change for them? Is it possible to make the change backward compatible? If we are not going to support using the new config going forward, I think it makes sense to at least issue a deprecation warning or guide user to use a new config.

carmmmm replied:
This update shouldn’t break anything for folks with an existing .synapseConfig. The file is still in the same location (~/.synapseConfig), and if someone’s using the old [authentication] section, it will still be picked up just like before. I've tested with both new profiles, and a config file with only an [authentication] section based off the previous design. 

Based on previous PR advice, I streamlined the docs to only show the new multi-profile format to encourage moving in that direction, but everything’s still backward compatible under the hood. But happy to add a quick note to the docs if that’d be more helpful!

linglp replied:
Got it. Thank you for clarifying.

linglp commented on docs/tutorials/command_line_client.md:
````diff
@@ -349,16 +349,17 @@ synapse list [-h] [-r] [-l] [-m] syn123
 
 ```bash
-synapse config [-h]
+synapse config [-h] [--profile PROFILE_NAME]
````
I tried: `synapse config -h --profile default` and it didn't quite work. I didn't see help message for profile flag: 
```
(venv) lpeng@MacBookPro synapsePythonClient % synapse config -h --profile default
usage: synapse config [-h]

options:
  -h, --help  show this help message and exit
```

linglp commented on docs/tutorials/command_line_client.md:
````diff
@@ -475,14 +476,17 @@ Verify credentials can be used to login to Synapse.
 
 ```bash
-synapse login [-h] [-u SYNAPSEUSER] [-p SYNAPSE_AUTH_TOKEN]
+synapse login [-h] [-u SYNAPSEUSER] [-p SYNAPSE_AUTH_TOKEN] [--profile PROFILE_NAME]
````
when I did `synapse login -h`, I expect to see explanation of flag "profile". But currently I am only seeing: 

```
usage: synapse login [-h] [-u SYNAPSEUSER] [-p SYNAPSE_AUTH_TOKEN]

options:
  -h, --help            show this help message and exit
  -u SYNAPSEUSER, --username SYNAPSEUSER
                        Username used to connect to Synapse
  -p SYNAPSE_AUTH_TOKEN, --password SYNAPSE_AUTH_TOKEN
                        Personal Access Token (aka: Synapse Auth Token) used to connect to Synapse.

```

carmmmm replied:
Thanks for pointing that out!  I had mistakenly added --profile as a top-level flag.

thomasyu888 replied:
One thing to my team, is should this be a top level or login level flag?

The reason I ask is for every command, you could technically specify a "profile"

BryanFauble replied:
IMO it should be a top level flag (And maybe also a command level flag to mirror how the `-p` flag?). With every command you should be able to use a specific profile during the execution of that command.

As far as it showing up when you run `synapse login -h`: It would be a "Nice to have" if it could show up during this help command, but, I don't know how enough knowledge on our CLI implementation to know if this is possible.


(This is on a branch without the changes from this PR)
FWIW, if you do `synapse -h` the `-p SYNAPSE_AUTH_TOKEN` shows up:

![image](https://github.com/user-attachments/assets/a42322c3-ad36-4709-8abd-42d0de898aa8)


It also shows up when you do `synapse login -h`

![image](https://github.com/user-attachments/assets/6bee8532-65bd-4e9b-ae30-0c098765a20d)


So this is telling me that we can likely do the same thing that the `-p` or `--password` flag is doing for this `--profile` flag

carmmmm replied:
Thanks all for the feedback — super helpful!

I've now updated the implementation so that --profile is supported both as a top-level flag (for use with any command) and as a login-level flag, similar to how --password is handled. 

Also made sure it now shows up correctly when you run synapse login -h, so users will see the description for the --profile flag alongside the other options.

Let me know if you think the placement or wording of the help message could be improved!

synapse -login -h
![Screenshot 2025-04-10 at 2 00 42 PM](https://github.com/user-attachments/assets/a27dd57b-24ab-4ffe-ac11-cc3f7fe78944)

synapse -h 

![Screenshot 2025-04-10 at 2 01 59 PM](https://github.com/user-attachments/assets/2351840b-076b-4972-9c12-28dbe4732be6)

linglp commented on docs/tutorials/command_line_client.md:
```diff
@@ -475,14 +476,17 @@ Verify credentials can be used to login to Synapse.
+|------------------|---------|----------------------------------------------------------------------------|-----------------------|
+| `-u, --username` | Named  | Username used to connect to Synapse.                                       |                       |
+| `-p, --password` | Named  | Synapse Auth Token (aka: Personal Access Token) used to connect to Synapse |                       |
+| `--profile`      | Named  | Name of the Synapse profile (from .synapseConfig) to log in under          |                       |
```
for consistency, should `--profile` also have a short cut?

thomasyu888 replied:
I personally think it's ok if it doesn't.

carmmmm replied:
I went ahead and added -r as a shortcut for --profile to keep it consistent with how -u and -p work for --username and --password. Totally open to removing it if folks feel strongly, but figured it might be convenient and clearer for users who are already used to the other flags having both short and long forms.

linglp commented on docs/tutorials/authentication.md:
```diff
@@ -33,47 +34,83 @@ Welcome, First Last!
+#For additional profiles
+> synapse --profile $MY_PROFILE_NAME config
+Synapse username (Optional): $MY_USERNAME
 
```
I tried adding a new profile using the command here, and I got message: 
```
[ERROR] Failed to perform command: <function config at 0x107d7edd0>
Traceback (most recent call last):
  File "/Users/lpeng/Documents/synapsePythonClient/venv/lib/python3.10/site-packages/synapseclient/__main__.py", line 1807, in perform_main
    args.func(args, syn)
  File "/usr/local/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/Users/lpeng/Documents/synapsePythonClient/venv/lib/python3.10/site-packages/synapseclient/__main__.py", line 591, in config
    with open(args.configPath, "w") as config_o:
PermissionError: [Errno 13] Permission denied: '/Users/lpeng/.synapseConfig'
```
Can we guide user in the tutorial so that this message can be avoided? I assume user needs to grant permission for .synapseConfig to be modified.

carmmmm replied:
Smart! Thank you!

linglp commented on docs/tutorials/command_line_client.md:
```diff
@@ -475,14 +476,17 @@ Verify credentials can be used to login to Synapse.
+| `--profile`      | Named  | Name of the Synapse profile (from .synapseConfig) to log in under          |                       |
 
+If --profile is provided, the credentials will be validated and the login will be associated with the given profile.
+If no profile is specified, the default '[default]' section in ~/.synapseConfig will be used.
```
if a duplicated profile name is used in .synapseConfig, what would happen?

linglp reviewed (commented):
thank you! I left some comments too

SageGJ commented on synapseclient/api/configuration_services.py:
```diff
@@ -79,16 +79,29 @@ def get_client_authenticated_s3_profile(
         The authentication section of the configuration file
     """
+
+    section = f"profile {profile}" if profile != "default" else profile
+
+    section_for_profile = get_config_section_dict(
+        section_name=section,
+        config_path=config_path,
+    )
+
+    if section_for_profile:
+        return section_for_profile
```
Unsure if this is what was intended in the last discussion, but could we also include a fallback to check for the old `authentication` section in cases where `section_for_profile is None`?

SageGJ commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -252,8 +269,16 @@ def get_credentials(
             [SynapseCredentials][synapseclient.core.credentials.cred_data.SynapseCredentials]
                 returned by the first non-None provider in its list, None otherwise
         """
+        selected_profile = user_login_args.profile or os.getenv("SYNAPSE_PROFILE")
```
In cases where the argument provided profile and the environment variable profile differ, the argument profile will be used. Should we also include a check that raises a message if the two specifications are not the same, so that users don't forget about one specification or the other and assume they're logged in under a profile they are not?

carmmmm replied:
Good idea! Thanks!

BWMac commented on docs/tutorials/authentication.md:
````diff
@@ -33,47 +34,83 @@ Welcome, First Last!
 ```
 
-## Use `.synapseConfig`
+## Use `.synapseConfig` for Multiple Profiles
````
I think the documentation does a good job describing how to use the `.synapseConfig` file, but I think the title was fine as it was since we can assume that the basic use case will still be users with only one profile. If anything, we should have a subsection within this one that describes how to go from using one profile to using multiple.
```suggestion
## Use `.synapseConfig`
```

BWMac commented on docs/tutorials/authentication.md:
```diff
@@ -33,47 +34,83 @@ Welcome, First Last!
-authtoken = <authtoken>
+[default]
+username = default_user
+#authtoken = default_auth_token
```
should this be uncommented too?

BWMac commented on docs/tutorials/authentication.md:
````diff
@@ -33,47 +34,83 @@ Welcome, First Last!
 syn = synapseclient.login()
 ```
 
+If logging in specifying the profile, it will log in with said profile:
````
```suggestion
Or, if you are logging into a specific profile, simply pass the profile name as an argument to `login()`: 
```

BWMac commented on synapseclient/__main__.py:
```diff
@@ -513,18 +513,24 @@ def _replace_existing_config(path, auth_section):
         config_text = config_o.read()
 
+    section_name = (
+        "default" if profile_name == "default" else f"profile {profile_name}"
```
NIT: No need to hard-code "default"
```suggestion
        profile_name if profile_name == "default" else f"profile {profile_name}"
```

BWMac commented on synapseclient/__main__.py:
```diff
@@ -534,21 +540,27 @@ def _replace_existing_config(path, auth_section):
+            flags=re.MULTILINE | re.DOTALL,
+        )
+
+    if profile_name:
```
NIT: I think this would be a little more readable if you just used `else` for the second condition since these are mutually exclusive paths in the code. Usually when I see this pattern I assume that it's possible for both code blocks to be executed.

BWMac commented on synapseclient/__main__.py:
```diff
@@ -564,15 +576,17 @@ def config(args, syn):
 
-    auth_section = "[authentication]\n"
+    profile_name = args.profile or "default"
+    section_header = f"[{profile_name}]\n" if profile_name == "default" else f"[profile {profile_name}]\n"
+    auth_section = section_header
```
Why not just assign this to `auth_section` initially?

carmmmm replied:
Ah I could've done that better, thanks!

BWMac commented on synapseclient/api/configuration_services.py:
```diff
@@ -79,16 +79,29 @@ def get_client_authenticated_s3_profile(
         The authentication section of the configuration file
     """
+
+    section = f"profile {profile}" if profile != "default" else profile
```
In other places you have done this the opposite way, for consistency you might want to stick with something like
```
section = profile if profile == "default" else f"profile {profile}"
```

BWMac commented on synapseclient/client.py:
```diff
@@ -231,7 +231,7 @@ def login(*args, **kwargs):
 
             import synapseclient
-            syn = synapseclient.login()
+            syn = synapseclient.login(profile="user1)
```
Rather than changing the first example, you should instead include a second example of logging in with a specified (non-default) profile since the base use case will still be the default behavior.

BWMac commented on .gitignore:
```diff
@@ -32,3 +32,5 @@ coverage.xml
 
 .ipynb_checkpoints
 *.ipynb
+synapseclient/.synapseConfig
```
Should this be included here if we are still going with the `.synapseConfig` file in the home directory being used? This serves as more of an example, no?

carmmmm replied:
I know we’re treating this as a template, but I added .synapseConfig to .gitignore just to prevent someone from accidentally filling it out with real credentials and pushing it. Happy to remove it if that feels unnecessary, just wanted to err on the side of caution!

BWMac replied:
I think we should not have this in the gitignore because if we wanted to change the structure of the config file we couldn't update the example without removing this anyway

BWMac commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -109,28 +110,40 @@ class UserArgsCredentialsProvider(SynapseCredentialsProvider):
-    ) -> Tuple[Union[str, None], Union[str, None]]:
-        config_dict = get_config_authentication(config_path=syn.configPath)
-        # check to make sure we didn't accidentally provide the wrong user
+        self, syn: "Synapse", user_login_args
```
NIT: missing type hint for `user_login_args`

BWMac commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -144,6 +157,10 @@ def _get_auth_info(
                 "in the `~/.synapseConfig` file.",
             )
 
+        # If authToken is explicitly provided, return it directly and skip profiles
+        if user_login_args.auth_token:
+            return user_login_args.username, user_login_args.auth_token
+
```
If we are going to provide this check, can we just do it right away so no other logic needs to be executed?

BWMac commented on synapseclient/core/credentials/credential_provider.py:
```diff
@@ -252,8 +269,16 @@ def get_credentials(
-            creds = provider.get_synapse_credentials(syn, user_login_args)
+            creds = provider.get_synapse_credentials(
+                syn,
+                UserLoginArgs(
+                    username=user_login_args.username,
+                    auth_token=user_login_args.auth_token,
+                    profile=selected_profile,
+                ),
```
`user_login_args` is already of type `UserLoginArgs` right? So can we just pass `user_login_args` straight in?

BWMac reviewed (commented):
Hi @carmmmm thank you so much for your contribution, this is awesome to see!

BWMac commented on tests/unit/synapseclient/unit_test_commandline.py:
```diff
@@ -593,16 +603,89 @@ def test_config_replace(
+
+    expected_auth_section = "[profile devprofile]\nusername=testuser\nauthtoken=authtoken123\n\n"
+    cmdline.config(args, syn)
+    mock_generate_new_config.assert_called_once_with(expected_auth_section, "devprofile")
```
We should include assertions for everything that we expect to be called and also things we do not expect to be called. For example, we are currently missing an expected call of `_authenticate_login` and an assertion saying that we expect `_replace_existing_config` not to be called.

BWMac commented on tests/unit/synapseclient/unit_test_commandline.py:
```diff
@@ -593,16 +603,89 @@ def test_config_replace(
+
+    expected_auth_section = "[profile prod]\nusername=testuser\nauthtoken=authtoken123\n\n"
+    cmdline.config(args, syn)
+    mock_replace_existing_config.assert_called_once_with(config_path, expected_auth_section, "prod")
```
Same comment as above.

carmmmm commented:
I’ve made the requested changes and addressed all the feedback. I’ve also added the missing tests and updated the code as discussed. Please review the updates when you get a chance, and let me know if anything needs to be modified. Thanks!

thomasyu888 commented:
Hmmm, this is a weird quirk to note

```
# this first one doesn't actually work
synapse -r tyu login
Synapse username (Optional): 
Auth token:

# but this works
synapse login -r tyu
```

But it's looking good! 

```
synapse -h
usage: synapse [-h] [--version] [-u SYNAPSEUSER] [-p SYNAPSE_AUTH_TOKEN]
               [-r PROFILE] [-c CONFIGPATH] [--debug] [--silent] [-s]
               [--otel {console,otlp}]
               {get,manifest,sync,store,add,mv,cp,get-download-list,associate,delete,query,submit,show,cat,list,config,set-provenance,get-provenance,set-annotations,get-annotations,create,store-table,onweb,login,test-encoding,get-sts-token,migrate}

Interfaces with the Synapse repository.

options:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  -u SYNAPSEUSER, --username SYNAPSEUSER
                        Username used to connect to Synapse
  -p SYNAPSE_AUTH_TOKEN, --password SYNAPSE_AUTH_TOKEN
                        Personal Access Token (aka: Synapse Auth Token) used to
                        connect to Synapse.
  -r PROFILE, --profile PROFILE
                        Name of the Synapse profile to use (from
                        ~/.synapseConfig). Defaults to 'default'.
```

BryanFauble commented:
@carmmmm @thomasyu888 @BWMac @SageGJ @linglp 

I created an additional pull request in the fork to address some items that I found while going through an manually testing the changes. Please review it here:

https://github.com/carmmmm/synapsePythonClient/pull/1

@carmmmm When we get approval please merge that PR into your fork. The other thing is that I didn't do anything to write up unit/integration tests to verify the subtle changes that I had made. If this is something you'd be willing to make sure work and cover the new logic that would be amazing!

BryanFauble commented:
> Hmmm, this is a weird quirk to note
> 
> ```
> # this first one doesn't actually work
> synapse -r tyu login
> Synapse username (Optional): 
> Auth token:
> 
> # but this works
> synapse login -r tyu
> ```
> 
> But it's looking good!
> 
> ```
> synapse -h
> usage: synapse [-h] [--version] [-u SYNAPSEUSER] [-p SYNAPSE_AUTH_TOKEN]
>                [-r PROFILE] [-c CONFIGPATH] [--debug] [--silent] [-s]
>                [--otel {console,otlp}]
>                {get,manifest,sync,store,add,mv,cp,get-download-list,associate,delete,query,submit,show,cat,list,config,set-provenance,get-provenance,set-annotations,get-annotations,create,store-table,onweb,login,test-encoding,get-sts-token,migrate}
> 
> Interfaces with the Synapse repository.
> 
> options:
>   -h, --help            show this help message and exit
>   --version             show program's version number and exit
>   -u SYNAPSEUSER, --username SYNAPSEUSER
>                         Username used to connect to Synapse
>   -p SYNAPSE_AUTH_TOKEN, --password SYNAPSE_AUTH_TOKEN
>                         Personal Access Token (aka: Synapse Auth Token) used to
>                         connect to Synapse.
>   -r PROFILE, --profile PROFILE
>                         Name of the Synapse profile to use (from
>                         ~/.synapseConfig). Defaults to 'default'.
> ```

@thomasyu888 https://github.com/carmmmm/synapsePythonClient/pull/1 patches this behavior. I had noticed as well. I have this note in the pull request description:

> I removed the support in the command line client to do synapse config --profile <PROFILE_NAME>, instead all commands need to specify the --profile flag first, like: synapse --profile <PROFILE_NAME> config. The reason for this change is because when flags were in both the root command and the sub-command it was causing data to be wiped out ie None in the Python string. Keeping it in a single spot provided consistency across the board. We can look to expand usage later on if this is problematic.

thomasyu888 commented:
@carmmmm / @BryanFauble not sure what's going on but see this:

```
(synapse-dev) ➜  synapsePythonClient git:(cb6b4c80) ✗ synapse login -r tyu
Welcome, Thomas Yu! You are using the 'tyu' profile.
Logged in as: thomas.yu (3324230)

Use `synapse config` to create or modify a Synapse configuration file. This will allow all commands to authenticate without passing in an Auth Token.
`synapse login` is only used to verify your credentials are valid to login.
(synapse-dev) ➜  synapsePythonClient git:(cb6b4c80) ✗ synapse -r tyu login
Welcome, Thomas Yu!

Logged in as: thomas.yu (3324230)

Use `synapse config` to create or modify a Synapse configuration file. This will allow all commands to authenticate without passing in an Auth Token.
`synapse login` is only used to verify your credentials are valid to login.
```

thomasyu888 reviewed (approved):
🔥 Thanks so much for your contribution @carmmmm !  heres's some tests that I did that worked

1. Having [authentication] and [default] and [profile] sections will pull from the default
2. Having [authentication] and [profile] but no [default] will pull from [authentication] if profile isn't specified
3. Having [authentication] and [profile] but no [default] will pull from [profile] if profile is specified
4. Having just [profile] but not specifying a profile will prompt the user to login
5. Having just [profile] but specifying a profile will work
6. Having only [authentication] still works

The `-r` parameter was removed from `synapse login` which is expected

linglp reviewed (commented):
Thanks for the hard work here. I tried using the config to log in again: 
```
[default]
username = linglp0122
authtoken = my_real_auth_token

[profile service-dpe]
username = synapse-service-dpe-team
authtoken = my_false_token 
```
I saw the following message when I ran the command: 
```
(venv) lpeng@MacBookPro synapsePythonClient % synapse --profile service-dpe login
Welcome, Lingling Peng! You are using the 'service-dpe' profile.
Logged in as: linglp0122 (3443707)

Use `synapse config` to create or modify a Synapse configuration file. This will allow all commands to authenticate without passing in an Auth Token.
`synapse login` is only used to verify your credentials are valid to login.
```
I think as a user, this message is a bit confusing because it seems like profile`synapse-service-dpe-team` was used but I actually got logged in using my default profile.

carmmmm commented:
@linglp Thank you for your feedback! I'm having trouble replicating your results if you could help clarify.  

If service-dpe uses a false token, then `synapse --profile service-dpe login` shouldn't work because you have an incorrect token associated with the login, but it also shouldn't login with the default.

I copied your synapseConfig example, 

```
[default]
username = monterocarmen
authtoken = my_real_auth_token

[profile service-dpe]
username = synapse-service-dpe-team
authtoken = An incorrect token 

```

but when i run `synapse --profile service-dpe login` I get this in return : 

```
(synapsePythonClient) (base) Carmens-MacBook-Pro-2:synapsePythonClient CarmenLaptop$ synapse --profile service-dpe login
[ERROR] Failed to perform command: <function login at 0x10b665b20>
Traceback (most recent call last):
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/client.py", line 6151, in _handle_synapse_http_error
    exceptions._raise_for_status(response, verbose=self.debug)
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/core/exceptions.py", line 194, in _raise_for_status
    raise SynapseHTTPError(message, response=response)
synapseclient.core.exceptions.SynapseHTTPError: 401 Client Error: 
Invalid access token

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/__main__.py", line 1811, in perform_main
    args.func(args, syn)
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/__main__.py", line 679, in login
    login_with_prompt(
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/__main__.py", line 1834, in login_with_prompt
    _authenticate_login(
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/__main__.py", line 1916, in _authenticate_login
    raise first_auth_ex or SynapseNoCredentialsError()
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/__main__.py", line 1902, in _authenticate_login
    syn.login(user, **login_kwargs_with_secret)
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/client.py", line 830, in login
    self.credentials = credential_provider_chain.get_credentials(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/core/credentials/credential_provider.py", line 309, in get_credentials
    creds = provider.get_synapse_credentials(syn, user_login_args)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/core/credentials/credential_provider.py", line 77, in get_synapse_credentials
    return self._create_synapse_credential(syn, *auth_info)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/core/credentials/credential_provider.py", line 88, in _create_synapse_credential
    profile = syn.restGET("/userProfile", auth=credentials)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/client.py", line 6262, in restGET
    response = self._rest_call(
               ^^^^^^^^^^^^^^^^
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/client.py", line 6237, in _rest_call
    self._handle_synapse_http_error(response)
  File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/client.py", line 6158, in _handle_synapse_http_error
    raise SynapseAuthenticationError(
synapseclient.core.exceptions.SynapseAuthenticationError: You are not logged in and do not have access to a requested resource.
```
And then if I run synapse` synapse --profile default login `or ` synapse login ` I get: 

```
(synapsePythonClient) (base) Carmens-MacBook-Pro-2:synapsePythonClient CarmenLaptop$ synapse login
Welcome, monterocarmen! You are using the 'default' profile.
Logged in as: monterocarmen (3534868)

Use `synapse config` to create or modify a Synapse configuration file. This will allow all commands to authenticate without passing in an Auth Token.
`synapse login` is only used to verify your credentials are valid to login.
```
 
Which is the expected response. But it seems we are getting two different responses, so if you could let me know where we differ! Thank you!

BryanFauble commented on docs/tutorials/command_line_client.md:
````diff
@@ -17,6 +17,7 @@ For help on specific commands, type:
 ```bash
 synapse [command] -h
 ```
+**Note:** All commands support the --profile (or -r) flag to specify which Synapse profile to use from your ~/.synapseConfig file. If not specified, the default profile is used.
````
This note is correct. 


`The -r parameter was removed from synapse login which is expected` is correct, BUT, the `-r` parameter is available on all commands, it's just not a flag on the sub commands.

Works:

`synapse -r <PROFILE> login`

Does not work:
`synapse login -r <PROFILE>`


Maybe there is better wording that could be used in this Note, but some form of it could be kept

carmmmm replied:
Ah, glad you mentioned this! I was actually removing the -r flag from the CLI to align with this change. Since neither -r nor --profile are listed as subcommands, and there is a dedicated section for subcommands, I don't think this will cause confusion. However, I can understand how, based on our previous experience, it might feel confusing. We could refer to "higher-level commands" versus "subcommands," but I'm not sure if that adds much clarity.

linglp replied:
oh very interesting. I think maybe clarify that will be helpful

linglp commented:
@carmmmm Sorry. I think I ran into this because I didn't install from the source. I could see the expected error message now. Thank you

BryanFauble commented:
> @linglp Thank you for your feedback! I'm having trouble replicating your results if you could help clarify.
> 
> If service-dpe uses a false token, then `synapse --profile service-dpe login` shouldn't work because you have an incorrect token associated with the login, but it also shouldn't login with the default.
> 
> I copied your synapseConfig example,
> 
> ```
> [default]
> username = monterocarmen
> authtoken = my_real_auth_token
> 
> [profile service-dpe]
> username = synapse-service-dpe-team
> authtoken = An incorrect token 
> ```
> 
> but when i run `synapse --profile service-dpe login` I get this in return :
> 
> ```
> (synapsePythonClient) (base) Carmens-MacBook-Pro-2:synapsePythonClient CarmenLaptop$ synapse --profile service-dpe login
> [ERROR] Failed to perform command: <function login at 0x10b665b20>
> Traceback (most recent call last):
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/client.py", line 6151, in _handle_synapse_http_error
>     exceptions._raise_for_status(response, verbose=self.debug)
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/core/exceptions.py", line 194, in _raise_for_status
>     raise SynapseHTTPError(message, response=response)
> synapseclient.core.exceptions.SynapseHTTPError: 401 Client Error: 
> Invalid access token
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/__main__.py", line 1811, in perform_main
>     args.func(args, syn)
>   File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py", line 81, in inner
>     return func(*args, **kwds)
>            ^^^^^^^^^^^^^^^^^^^
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/__main__.py", line 679, in login
>     login_with_prompt(
>   File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py", line 81, in inner
>     return func(*args, **kwds)
>            ^^^^^^^^^^^^^^^^^^^
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/__main__.py", line 1834, in login_with_prompt
>     _authenticate_login(
>   File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py", line 81, in inner
>     return func(*args, **kwds)
>            ^^^^^^^^^^^^^^^^^^^
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/__main__.py", line 1916, in _authenticate_login
>     raise first_auth_ex or SynapseNoCredentialsError()
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/__main__.py", line 1902, in _authenticate_login
>     syn.login(user, **login_kwargs_with_secret)
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/client.py", line 830, in login
>     self.credentials = credential_provider_chain.get_credentials(
>                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/core/credentials/credential_provider.py", line 309, in get_credentials
>     creds = provider.get_synapse_credentials(syn, user_login_args)
>             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/core/credentials/credential_provider.py", line 77, in get_synapse_credentials
>     return self._create_synapse_credential(syn, *auth_info)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/core/credentials/credential_provider.py", line 88, in _create_synapse_credential
>     profile = syn.restGET("/userProfile", auth=credentials)
>               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/client.py", line 6262, in restGET
>     response = self._rest_call(
>                ^^^^^^^^^^^^^^^^
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/client.py", line 6237, in _rest_call
>     self._handle_synapse_http_error(response)
>   File "/Users/CarmenLaptop/Desktop/coding/synapsePythonClient/synapseclient/client.py", line 6158, in _handle_synapse_http_error
>     raise SynapseAuthenticationError(
> synapseclient.core.exceptions.SynapseAuthenticationError: You are not logged in and do not have access to a requested resource.
> ```
> 
> And then if I run synapse`synapse --profile default login`or `synapse login` I get:
> 
> ```
> (synapsePythonClient) (base) Carmens-MacBook-Pro-2:synapsePythonClient CarmenLaptop$ synapse login
> Welcome, monterocarmen! You are using the 'default' profile.
> Logged in as: monterocarmen (3534868)
> 
> Use `synapse config` to create or modify a Synapse configuration file. This will allow all commands to authenticate without passing in an Auth Token.
> `synapse login` is only used to verify your credentials are valid to login.
> ```
> 
> Which is the expected response. But it seems we are getting two different responses, so if you could let me know where we differ! Thank you!

@carmmmm I talked with @linglp and this is not an issue with the code. It was an issue where the latest commits were not pulled in

carmmmm commented:
@BryanFauble You likely already know this, but just in case — I wanted to flag that since _get_auth_info() is an abstract method, changing its return from a 2-value to a 3-value tuple technically updates the method contract and could break any external subclass still returning only two values.

It might be worth noting in the release notes that any custom credential provider subclasses should now return three values. That said, if external subclasses are rare or unlikely, maybe it's not an issue — I'm not sure how commonly this method is subclassed outside the repo.

If external subclassing is a bigger concern, another option (totally optional!) could be returning a namedtuple instead for added safety and clarity when unpacking.

BryanFauble commented:
> @BryanFauble You likely already know this, but just in case — I wanted to flag that since _get_auth_info() is an abstract method, changing its return from a 2-value to a 3-value tuple technically updates the method contract and could break any external subclass still returning only two values.
> 
> It might be worth noting in the release notes that any custom credential provider subclasses should now return three values. That said, if external subclasses are rare or unlikely, maybe it's not an issue — I'm not sure how commonly this method is subclassed outside the repo.
> 
> If external subclassing is a bigger concern, another option (totally optional!) could be returning a namedtuple instead for added safety and clarity when unpacking.

@carmmmm Thanks for bringing attention to this.

items in the `core` have this notice:
> This section is for super users / developers only. These functions are subject to change as they are internal development functions. Use at your own risk.

https://python-docs.synapse.org/en/stable/reference/core/


I don't have concerns that we include breaking changes here.

carmmmm commented:
Just added the tests! Let me know if there’s anything else you wanted covered.

Also made a small logic fix in __main__.py—one of the tests wasn’t passing because the previous logic was allowing [authentication] to be added twice instead of replacing it properly.

BryanFauble reviewed (approved):
Thanks for all these changes. I am going to merge this into a feature branch in the main repo so that integration tests can run in the regular github pipeline. If all is well ill plan to merge this to develop tomorrow

carmmmm commented:
Hey @BryanFauble thanks for merging! I noticed a few pre-commit hooks ran and made formatting changes after the PR was closed. Would it be helpful if I opened a quick follow-up PR with those cleanup changes? Sorry about that!

BryanFauble commented:
> Hey @BryanFauble thanks for merging! I noticed a few pre-commit hooks ran and made formatting changes after the PR was closed. Would it be helpful if I opened a quick follow-up PR with those cleanup changes? Sorry about that!

@carmmmm thanks, I already took care of it. New run: https://github.com/Sage-Bionetworks/synapsePythonClient/actions/runs/14605880964

New PR: https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1194
